cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1202 @@
+	requestedStores map[string]*requestedStore
+
+	totalSpans struct {
+		sync.RWMutex
+		m *spanz.HashMap[SubscriptionID]
+		v map[SubscriptionID]*requestedTable
+	}
+}
+
+type resolveLockTask struct {
+	regionID   uint64
+	maxVersion uint64
+	state      *regionlock.LockedRange
+	enter      time.Time
+}
+
+type rangeTask struct {
+	span           tablepb.Span
+	requestedTable *requestedTable
+}
+
+type requestedStore struct {
+	storeID    uint64
+	storeAddr  string
+	nextStream atomic.Uint32
+	streams    []*requestedStream
+}
+
+type requestedStream struct {
+	streamID uint64
+
+	// To trigger a connect action lazily.
+	preFetchForConnecting *singleRegionInfo
+	requests              *chann.DrainableChann[singleRegionInfo]
+
+	// init is for initializing client and conn asynchronously.
+	init   func(ctx context.Context) error
+	client cdcpb.ChangeData_EventFeedClient
+	conn   *sharedConn
+
+	requestedRegions struct {
+		sync.RWMutex
+		// map[subscriptionID]map[regionID]*regionFeedState
+		m map[SubscriptionID]map[uint64]*regionFeedState
+	}
+}
+
+type requestedTable struct {
+	subscriptionID SubscriptionID
+
+	span      tablepb.Span
+	startTs   model.Ts
+	rangeLock *regionlock.RegionRangeLock
+	eventCh   chan<- MultiplexingEvent
+
+	// To handle table removing.
+	stopped atomic.Bool
+
+	// To handle lock resolvings.
+	postUpdateRegionResolvedTs func(regionID uint64, state *regionlock.LockedRange)
+	staleLocksVersion          atomic.Uint64
+}
+
+// NewSharedClient creates a client.
+func NewSharedClient(
+	changefeed model.ChangeFeedID,
+	cfg *config.KVClientConfig,
+	filterLoop bool,
+	pd pd.Client,
+	grpcPool GrpcPool,
+	regionCache *tikv.RegionCache,
+	pdClock pdutil.Clock,
+	lockResolver txnutil.LockResolver,
+) *SharedClient {
+	s := &SharedClient{
+		changefeed: changefeed,
+		config:     cfg,
+		clusterID:  0,
+		filterLoop: filterLoop,
+
+		pd:           pd,
+		grpcPool:     grpcPool,
+		regionCache:  regionCache,
+		pdClock:      pdClock,
+		lockResolver: lockResolver,
+
+		requestRangeCh: chann.NewAutoDrainChann[rangeTask](),
+		regionCh:       chann.NewAutoDrainChann[singleRegionInfo](),
+		regionRouter:   chann.NewAutoDrainChann[singleRegionInfo](),
+		resolveLockCh:  chann.NewAutoDrainChann[resolveLockTask](),
+		errCh:          chann.NewAutoDrainChann[regionErrorInfo](),
+
+		requestedStores: make(map[string]*requestedStore),
+	}
+	s.totalSpans.m = spanz.NewHashMap[SubscriptionID]()
+	s.totalSpans.v = make(map[SubscriptionID]*requestedTable)
+	s.initMetrics()
+	return s
+}
+
+// AllocSubscriptionID gets an ID can be used in `Subscribe`.
+func (s *SharedClient) AllocSubscriptionID() SubscriptionID {
+	return SubscriptionID(subscriptionIDGen.Add(1))
+}
+
+// Subscribe the given table span.
+// NOTE: `span.TableID` must be set correctly.
+func (s *SharedClient) Subscribe(
+	subID SubscriptionID, span tablepb.Span, startTs uint64,
+	eventCh chan<- MultiplexingEvent,
+) (existSubID SubscriptionID, success bool) {
```

Comment:
The SubID returned by the function should be used, which avoids calling AllocSubscriptionID externally.

Reply:
It's better to call `AllocSubscriptionID` before `Subscribe` so that `MultiplexingPuller` can install some handlers for the allocatedID before `Subscribe`.